### ISSUE TITLE
How can we test a URL with different params concurrently

### ISSUE THREAD

**Issue by vivekbrahmadandi** [alexfernandez/loadtest#111]
I have a load testing use case where url is same but params are different it should be tested concurrently and need responses . Is it can be done using loadtest

**Comment by alexfernandez** [comment]
Hi, what kind of params are different? Can you show a couple of examples?

**Comment by ChrisCinelli** [comment]
I think in general it would be nice to have a file with different URLs and being able to loop through them

**Comment by alexfernandez** [comment]
Sounds interesting, ¿care to send a patch?

**Comment by sunojvijayan** [comment]
Hi I have to test with different parameters. Is it possible. This does not work but is something like this possible.


var loadtest = require('loadtest');

var no = 0;

function optionsObject() {
    no++;
    return {
        url: 'http://172.31.20.165:3000/api/v1/registeruser',
        maxRequests: 2,
        concurrency: 1,
        method: 'POST',
        contentType: 'application/json',
        body: {
            userId: "12323123" + no
            }
    }
}

loadtest.loadTest(optionsObject(), function (error, result) {
    if (error) {
        console.log('Got an error: %s', error);
    } else {
        console.log(result);
    }
});

Please help?

**Comment by alexfernandez** [comment]
It should not be too difficult. I am willing to include it if someone sends a pull request.

**Comment by sunojvijayan** [comment]
Hi, Any update on this please.

**Comment by alexfernandez** [comment]
@sunojvijayan There is nobody working on this. I suggest that you tackle it yourself. This is an open source project and we are all volunteers, it would be great if you could submit a patch so the community benefits.

**Comment by ffoysal** [comment]
@sunojvijayan I sent a pull request for your issue  to @alexfernandez once it is merged hope will help you

**Comment by alexfernandez** [comment]
Published @ffoysal's fix as 2.1.0 (thanks!), please reopen if you need anything else.

**Comment by sunojvijayan** [comment]
Can you please explain how to use different data?

**Comment by ffoysal** [comment]
It is explained under indexParam
Thanks

**Comment by sunojvijayan** [comment]
var loadtest = require('loadtest');

var no = 0;

var indexParam = {
        id_1: 12323123,
        id_2: 12323124
    }

function optionsObject() {
no++;
return {
url: 'http://url--------------/api/v1/registeruser',
maxRequests: 2,
concurrency: 1,
method: 'POST',
contentType: 'application/json',
body: {
userId: indexParam
}
}
}

loadtest.loadTest(optionsObject(), function (error, result) {
if (error) {
console.log('Got an error: %s', error);
} else {
console.log(result);
}
});


I tried this but got error. Please help

**Comment by alexfernandez** [comment]
You cannot choose the set of indexes that will be used by `indexParam`; you are just guaranteed to get an unique value each time. At the moment `loadtest` will choose them to be 1, 2, 3… in sequence.

**Comment by sunojvijayan** [comment]
Can you please show in my example how its done?

**Comment by alexfernandez** [comment]
Sure, you just need to change `optionsObject()`:

```
function optionsObject() {
  // next line is useless; this function only called once
  // no++;
  return {
    url: 'http://url--------------/api/v1/registeruser',
    maxRequests: 2,
    concurrency: 1,
    method: 'POST',
    contentType: 'application/json',
    // say what string will be replaced by the index
    indexParam: 'indexValue',
    body: {
      userId: 'user' + indexValue,
    },
  };
}
```

Now every time `indexValue` is found, either in the URL or in the body, it will be replaced by `1`, `2`…

**Comment by sunojvijayan** [comment]
var loadtest = require('loadtest');

function optionsObject() {
    //no++;

    return {
        url: 'http://172.31.20.165:3000/api/v1/registeruser',
        maxRequests: 2,
        concurrency: 1,
        method: 'POST',
        contentType: 'application/json',
        indexParam: 'indexValue',
        body: {
            user_id: 'indexValue'
        }
    }
}

loadtest.loadTest(optionsObject(), function (error, result) {
    if (error) {
        console.log('Got an error: %s', error);
    } else {
        console.log(result);
    }
});

Tried this, the first time and second time i am getting value as 0 only

**Comment by sunojvijayan** [comment]
Any idea why this is happening

**Comment by alexfernandez** [comment]
The remaining times are fine? If so, looks like a bug in loadtest.

**Comment by sunojvijayan** [comment]
No I mean I only ran it with 
maxRequests: 2,
 concurrency: 1,

Both times i got the value as 0

**Comment by alexfernandez** [comment]
Definitely a bug, reopening.

**Comment by ffoysal** [comment]
@sunojvijayan indexParam value depends on number of concurrency. as index always start at 0 so the you got always 0. if you increase concurrency: 2, you will get 0 and 1. the indexParam value does not depend on maxRequest.

**Comment by alexfernandez** [comment]
@ffoysal It works as you say, but I think it would be more useful if the `indexParam` was increased with each request, don't you think?

**Comment by sunojvijayan** [comment]
That would make it more useful. Don't you think.

**Comment by sunojvijayan** [comment]
If you can do it it would be really helpful for us. Thanks in advance.

**Comment by ffoysal** [comment]
suppose
url: 'http://172.31.20.165:3000/api/v1/registeruser',
    maxRequests: 4,
    concurrency: 2,
    method: 'POST',
    contentType: 'application/json',
    indexParam: 'indexValue',
    body: {
        user_id: 'indexValue'
    }


If we change the way it works now then which two url will run concurrently ??

and which urls will be used to run within a minute ???

**Comment by alexfernandez** [comment]
@ffoysal If it is changed, it would send:

    {"user_id": "0"}
    {"user_id": "1"}
    {"user_id": "2"}
    {"user_id": ...}

sequentially, with the `indexValue` increasing for every message. No two messages would be alike.

**Comment by shaishab** [comment]
If you want to use **request Index** value or **random** value in your request body then can use **requestGenerator** function in options.  Can use bellow code .

`function statusCallback(error, result, latency) {
	counter++;
	 console.log('Current latency %j, result %j, error %j', latency, result, error);
	console.log('----');
	console.log('Request elapsed milliseconds: ', result.requestElapsed);
	console.log('Request index: ', result.requestIndex);
	console.log('Request loadtest() instance index: ', result.instanceIndex);
}


var options = {
	url: 'http://localhost:3000/articles',
	concurrency: 2,
	maxRequests: 10,
	method: 'POST',
	contentType: 'application/json',
	headers: {
		"authToken": 'eyJ0eXAiOiJKV1QiLCJhbG'
	},
	requestIndex: 0,// set initial value to use in request body
	requestGenerator: function(params, options, client, callback)
	{
		// if you want to set random value can use bellow line instead of "user_id": params.requestIndex
		//"user_id": Math.random().toString(36).substring(7)
		var body = {
			"user_id": params.requestIndex
		};
		params.requestIndex += 1;

		var message = JSON.stringify(body);
		options.headers['Content-Length'] = message.length;
		options.headers['Content-Type'] = 'application/json';

		var request = client(options, callback);
		request.write(message);
		return request;
	},
	statusCallback: statusCallback
};

loadtest.loadTest(options, function (error, result) {
	if (error) {
		return console.error('Got an error: %s', error);
	}
	console.log('Tests run successfully');
});`

In my code I used **requestIdex** and initialize with 0 and in **requestGenerator** function used to set user-id like:  `"user_id": params.requestIndex`

**Comment by michaelsync** [comment]
Is there any way to use it with pre-defined values instead of 1,2,3... ?

**Comment by alexfernandez** [comment]
Closing due to inactivity, please reopen if needed.
